pkg/store/bucket.go

```diff
@@ -1522,40 +1628,45 @@ func (s *BucketStore) LabelValues(ctx context.Context, req *storepb.LabelValuesR
 				}
 				result = res
 			} else {
-				seriesSet, _, err := blockSeries(
-					newCtx,
-					b.extLset,
-					indexr,
-					nil,
+				seriesReq := &storepb.SeriesRequest{
+					MinTime:    req.Start,
+					MaxTime:    req.End,
+					SkipChunks: true,
+				}
+				blockClient := newBlockSeriesClient(newCtx, s.logger, b, seriesReq, nil, bytesLimiter, nil, true, SeriesBatchSize)
+
+				if err := blockClient.ExpandPostings(
 					reqSeriesMatchersNoExtLabels,
-					nil,
 					seriesLimiter,
-					bytesLimiter,
-					true,
-					req.Start,
-					req.End,
-					nil,
-					nil,
 					s.metrics.emptyPostingCount,
-					false,
-				)
-				if err != nil {
-					return errors.Wrapf(err, "fetch series for block %s", b.meta.ULID)
+				); err != nil {
+					return err
 				}
 
 				// Extract given label's value from all series and deduplicate them.
 				// We don't need to deal with external labels, since they are already added by blockSeries.
 				values := map[string]struct{}{}
-				for seriesSet.Next() {
-					ls, _ := seriesSet.At()
-					val := ls.Get(req.Label)
+				for {
+					ls, err := blockClient.Recv()
+					if err == io.EOF {
+						break
+					}
+					if err != nil {
+						return errors.Wrapf(err, "iterate series for block %s", b.meta.ULID)
+					}
+
+					if ls.GetWarning() != "" {
+						return errors.Wrapf(errors.New(ls.GetWarning()), "iterate series for block %s", b.meta.ULID)
```

Comment:
Shall we condier warnings as errors?

Reply:
We only do this for `LabelNames` and `LabelValues`. I don't think we emit warnings for these RPCs yet so I had to consider them as errors.